docs/rfcs/025-generation-numbers.md

```diff
@@ -0,0 +1,468 @@
+  lead them to the most recent writer's latest metadata.
+- Alternatively, whatever drives failover (perhaps the control plane) could also provide the hint to new
+  attachments of what the most recent active attachment was, so that the newly attached node already
+  knows the generation numbers of the node that last wrote metadata for this tenant.
+
+### Control Plane Changes
+
+#### Managing generation numbers
+
+The control plane is responsible for issuing generation numbers: this work is done in the control plane
+because without adding a consensus system to the pageservers, they do not have a way to reliably
+allocate unique generation numbers while tolerating network partitions.
+
+Assigning generation numbers to attachments is very simple, because designating a particular pageserver
+as the attachment point for a tenant is already a database transaction: this transaction just needs
+to increment the generation number as well as setting the pageserver.
+
+Assigning generation numbers to nodes at startup is a larger change: pageserver startup must be
+gated on communication from the control plane. We may either implement this via
+
+- pull: the pageserver calling up to the control plane to announce itself and request a generation number
+- push: in the opposite direction by having the control plane call down to the pageserver to give it
+  a generation number.
+
+The choice of pull or push does not affect the proposed storage format changes.
+
+#### Synchronizing attachments on pageserver startup
+
+In "pageserver startup changes" above, it is noted that the pageserver should synchronize its list of
+attachments on startup, to avoid starting redundant writes to S3 for tenants that are already being
+written from some other more up-to-date node.
+
+However, it is important to avoid O(tenant_count) communications where possible, to avoid excessive
+load on the control plane. Especially, there are two common special cases that can be made very
+fast:
+
+- A pageserver that restarts without any attachments being changed: this would happen if a pageserver
+  restarts quickly, before the control plane has even noticed an outage and moved any attachments.
+- A pageserver who has had _all_ their attachments removed between generations: this would happen
+  if the control plane has responded to the node's absence by re-attaching all tenants to
+  other nodes, before the original node returns to service.
+
+### Safekeeper changes
+
+As an optimization, the safekeeper may track the most recent known generation for each pageserver node ID,
+and respond with an error to any communications from older generations: this would help isolated
+pageserver nodes to shut down more quickly once they realize that they are stale, rather than continuing
+to put load on the safekeeper and write objects to S3 that are already superseded by later generations.
+
+For strict safety, the safekeeper should ensure that it does not trim any data based on messages
```

Comment:
This looks like a redundant requirement. If safekeeper received remote_consistent_lsn the data is durable on s3 regardless of which pageserver uploaded it. New pageserver wont delete data that was uploaded by previous PS right? In case we have multiple pageservers active at the same time it should be safe for SK to pick `max(remote_consistent_lsn)`